packages/block-editor/src/components/typewriter/index.js

```diff
@@ -0,0 +1,154 @@
+		if ( this.isSelectionEligibleForScroll() ) {
+			this.caretRect = computeCaretRect();
+		}
+	}
+
+	maintainCaretPositionOnSelectionChange() {
+		document.removeEventListener( 'selectionchange', this.maintainCaretPositionOnSelectionChange );
+		this.maintainCaretPosition();
+	}
+
+	computeCaretRectOnSelectionChange() {
+		document.removeEventListener( 'selectionchange', this.computeCaretRectOnSelectionChange );
+		this.computeCaretRect();
+	}
+
+	isSelectionEligibleForScroll() {
+		return (
+			// There should be one and only one block selected.
+			this.props.selectedBlockClientId &&
+			// The component must contain the selection.
+			this.ref.current.contains( document.activeElement ) &&
+			// The active element must be contenteditable.
+			document.activeElement.getAttribute( 'contenteditable' )
+		);
+	}
+
+	maintainCaretPosition() {
+		if ( ! this.caretRect ) {
+			return this.computeCaretRect();
+		}
+
+		if ( ! this.isSelectionEligibleForScroll() ) {
+			return;
+		}
+
+		const currentCaretRect = computeCaretRect();
+
+		if ( ! currentCaretRect ) {
+			return;
+		}
+
+		const diff = currentCaretRect.y - this.caretRect.y;
+
+		if ( diff === 0 ) {
+			return;
+		}
+
+		const scrollContainer = getScrollContainer( this.ref.current );
+
+		// The page must be scrollable.
+		if ( ! scrollContainer ) {
+			return;
+		}
+
+		// The scroll container may be different depending on the viewport
+		// width.
+		if ( scrollContainer === document.body ) {
+			window.scrollBy( 0, diff );
+		} else {
+			scrollContainer.scrollTop += diff;
+		}
+	}
+
+	onMouseDown() {
+		document.addEventListener( 'selectionchange', this.computeCaretRectOnSelectionChange );
+	}
+
+	onTouchStart() {
+		document.addEventListener( 'selectionchange', this.computeCaretRectOnSelectionChange );
+	}
+
+	onKeyDown() {
+		// Ensure the any remaining request is cancelled.
+		window.cancelAnimationFrame( this.rafId );
+		// Use an animation frame for a smooth result.
+		this.rafId = window.requestAnimationFrame( this.maintainCaretPosition );
+		// In rare cases, the selection is not updated before the animation
+		// frame. This selection change listener is a back up.
+		document.addEventListener( 'selectionchange', this.maintainCaretPositionOnSelectionChange );
+	}
+
+	render() {
+		// Disable reason: Wrapper itself is non-interactive, but must capture
+		// bubbling events from children to determine focus transition intents.
+		/* eslint-disable jsx-a11y/no-static-element-interactions */
+		return (
+			<div
+				ref={ this.ref }
+				onKeyDown={ this.onKeyDown }
+				onKeyUp={ this.onKeyUp }
+				onMouseDown={ this.onMouseDown }
+				onTouchStart={ this.onTouchStart }
+			>
+				{ this.props.children }
+			</div>
+		);
+		/* eslint-enable jsx-a11y/no-static-element-interactions */
+	}
+}
+
+/**
+ * Ensures that the text selection keeps the same vertical distance from the
+ * viewport during keyboard events within this component.
+ */
+export default compose( [
```

Comment:
`compose` isn't strictly necessary here being a single higher-order component.

```suggestion
export default withSelect( ( select ) => {
	const { getSelectedBlockClientId } = select( 'core/block-editor' );
	return { selectedBlockClientId: getSelectedBlockClientId() };
} )( TypeWriter );
```

Reply:
Aside: If you choose to implement this as a function component, we can consider `useSelect` hook instead.

https://github.com/WordPress/gutenberg/tree/master/packages/data#useSelect